sdk/python/kfp/compiler/compiler_test.py

```diff
@@ -3180,5 +3180,232 @@ def pipeline_component(input_list: Input[List[Artifact]]):
+        def print_op(message: str = 'default'):
+            print(message)
+
+        @dsl.pipeline()
+        def my_pipeline():
+            task1 = print_op()
+            task2 = print_op(message='different').set_display_name('name')
+
+        pipeline_spec = my_pipeline.pipeline_spec
+
+        self.assertEqual(len(pipeline_spec.components), 1)
+
+        # check one component spec and executor exist
+        self.assertIn('comp-print-op', pipeline_spec.components)
+        self.assertIn(
+            'exec-print-op', pipeline_spec.deployment_spec.fields['executors']
+            .struct_value.fields)
+        self.assertNotIn('comp-print-op-2', pipeline_spec.components)
+        self.assertNotIn(
+            'exec-print-op-2', pipeline_spec.deployment_spec.fields['executors']
+            .struct_value.fields)
+
+        # check tasks point to correct component spec
+
+        self.assertEqual(
+            pipeline_spec.root.dag.tasks['print-op'].component_ref.name,
+            'comp-print-op')
+        self.assertEqual(
+            pipeline_spec.root.dag.tasks['print-op-2'].component_ref.name,
+            'comp-print-op')
+
+        # check task specific information is saved
+        self.assertEqual(
+            pipeline_spec.root.dag.tasks['print-op-2'].inputs
+            .parameters['message'].runtime_value.constant.string_value,
+            'different')
+
+        self.assertEqual(
+            pipeline_spec.root.dag.tasks['print-op-2'].task_info.name, 'name')
+
+    def test_pipeline_in_pipeline_would_dedupe(self):
+
+        @dsl.component
+        def print_op(msg: str):
+            print(msg)
+
+        @dsl.pipeline(name='graph-component')
+        def graph_component(msg: str):
+            print_op(msg=msg)
+
+        @dsl.pipeline(name='test-pipeline')
+        def my_pipeline():
+            task1 = graph_component(msg='hello')
+            task2 = graph_component(msg='hello')
+
+        pipeline_spec = my_pipeline.pipeline_spec
+
+        self.assertEqual(len(pipeline_spec.components), 3)
+
+        # check both tasks component specs exist
+        self.assertIn('comp-graph-component', pipeline_spec.components)
+        self.assertIn('comp-graph-component-2', pipeline_spec.components)
+
+        # check only one executor exist
+        self.assertIn(
+            'exec-print-op', pipeline_spec.deployment_spec.fields['executors']
+            .struct_value.fields)
+        self.assertNotIn(
+            'exec-print-op-2', pipeline_spec.deployment_spec.fields['executors']
+            .struct_value.fields)
+
+    def test_clean_up_names(self):
+
+        @dsl.component
+        def print_op(message: str = 'default'):
+            print(message)
+
+        @dsl.pipeline()
+        def my_pipeline(pipeline_input: str = 'pipeine_input'):
+            task1 = print_op()
+            task2 = print_op()
+            task3 = print_op()
+            task4 = print_op().set_cpu_limit('4').set_memory_limit('14Gi')
```

Comment:
Is it possible there's a bug here? If I reorder as the following, the tests fail (non-deterministically):

```python
        @dsl.pipeline()
        def my_pipeline(pipeline_input: str = 'pipeine_input'):
            task1 = print_op()
            task4 = print_op().set_cpu_limit('4').set_memory_limit('14Gi')
            task2 = print_op()
            task3 = print_op()
```

Reply:
Good find! It is a bug that happens because the the `_clean_up_component_spec_names` function actually updates the name while still in the for loop that collects the names to be changed